### PR TITLE
Search PDF files with Khoj. Integrate with LangChain

### PR DESCRIPTION
**Pull request by debanjum**
- 18081b3 Use LangChain to call GPT for Khoj Chat
  - efcf7d1 Extract prompts for GPT used by Khoj Chat into LangChain prompt templates
  
- Allow Search (and Chat) on PDF with Khoj
  - 286b500 Create PDF to JSONL Processor: Convert PDF content into standardized JSONL format
  - acd14a5 Expose PDF search type via Khoj server API
  - Enable querying PDF files via Obsidian (bbe3bf9), Emacs (e389294) and Web (8514400) interfaces

### CODE REVIEW

File: src/khoj/processor/conversation/gpt.py [hunk 1]
```diff
@@ -102,86 +82,38 @@ def extract_questions(text, model="text-davinci-003", conversation_log={}, api_k
...
-Q: {text}
-
-"""
+    prompt = prompts.extract_questions.format(
+        current_date=today.strftime("%A, %Y-%m-%d"),
+        last_new_year=last_new_year.strftime("%Y"),
+        last_new_year_date=last_new_year.strftime("%Y-%m-%d"),
+        current_new_year_date=current_new_year.strftime("%Y-%m-%d"),
+        bob_tom_age_difference={current_new_year.year - 1984 - 30},
+        bob_age={current_new_year.year - 1984},
```
Comment (sabaimran) [inline src/khoj/processor/conversation/gpt.py:91]: What are these two values for?

File: src/khoj/processor/conversation/prompts.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+Answer (in second person):"""
+)
+
+
+## Extract Questions
+## --
+extract_questions = PromptTemplate.from_template(
+    """
+You are Khoj, an extremely smart and helpful search assistant with the ability to retrieve information from the users notes.
```
Comment (sabaimran) [inline src/khoj/processor/conversation/prompts.py:81]: nit

```suggestion
You are Khoj, an extremely smart and helpful search assistant with the ability to retrieve information from the user's notes.
```

File: src/khoj/processor/conversation/prompts.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+)
+
+
+## Extract Search Type
+## --
+search_type = """
+Objective: Extract search type from user query and return information as JSON
+
+Allowed search types are listed below:
+  - search-type=["notes","ledger","image","music"]
```
Comment (sabaimran) [inline src/khoj/processor/conversation/prompts.py:146]:
```suggestion
  - search-type=["notes","ledger","image","music","pdf"]
```

File: src/khoj/processor/conversation/prompts.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+A:{ "search-type": "music" }
+Q:How much did I spend at Subway for dinner last time?
+A:{ "search-type": "ledger" }
+Q:What was that popular Sri lankan song that Alex had mentioned?
+A:{ "search-type": "music" }
+Q:Can you recommend a movie to watch from my notes?
+A:{ "search-type": "notes" }
+Q:When did I buy Groceries last?
+A:{ "search-type": "ledger" }
+Q:When did I go surfing last?
```
Comment (sabaimran) [inline src/khoj/processor/conversation/prompts.py:161]:
```
Q: What did the lease say about early termination
A: { "search-type": "pdf" }
```

File: src/khoj/processor/pdf/pdf_to_jsonl.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+    @staticmethod
+    def convert_pdf_entries_to_maps(parsed_entries: List[str], entry_to_file_map) -> List[Entry]:
+        "Convert each PDF entries into a dictionary"
+        entries = []
+        for parsed_entry in parsed_entries:
+            entry_filename = Path(entry_to_file_map[parsed_entry])
+            # Append base filename to compiled entry for context to model
+            heading = f"{entry_filename.stem}\n"
+            compiled_entry = f"{heading}{parsed_entry}"
```
Comment (sabaimran) [inline src/khoj/processor/pdf/pdf_to_jsonl.py:114]: I believe this answers my above question with regards to the title rendered in search view.

**Review by sabaimran — APPROVED** [review]
Nice! This is awesome 🚀

File: pyproject.toml [hunk 6]
```diff
@@ -53,6 +54,8 @@ dependencies = [
     "torch == 1.13.1",
     "uvicorn == 0.17.6",
     "aiohttp == 3.8.4",
+    "langchain >= 0.0.187",
+    "pypdf >= 3.9.0",
```
[the remark this answers is not in the document]
Comment (debanjum) [reply pyproject.toml:58]: Yeah, I'd started with fixed dependency versions too but started to move towards looser dependency specifications. Then I read some stuff that recommended not trying to lock dependency to a specific version unless explicitly required to (I can't find the articles now!). Not bought too strongly on either side yet. But it avoids the dependency conflicts that people face if they've already installed the dependencies before, you get the fixes that the package developers make and you don't have to be as involved in the dependency management of your package.

File: src/interface/obsidian/src/search_modal.ts [hunk 7]
```diff
@@ -89,12 +89,24 @@ export class KhojSearchModal extends SuggestModal<SearchResult> {
...
         let encodedQuery = encodeURIComponent(query);
-        let searchUrl = `${this.setting.khojUrl}/api/search?q=${encodedQuery}&n=${this.setting.resultsCount}&r=${this.rerank}&t=markdown`;
-        let response = await request(searchUrl);
-        let data = JSON.parse(response);
-        let results = data
+        let searchUrl = `${this.setting.khojUrl}/api/search?q=${encodedQuery}&n=${this.setting.resultsCount}&r=${this.rerank}`;
+
+        // Get search results for markdown and pdf files
+        let mdResponse = await request(`${searchUrl}&t=markdown`);
+        let pdfResponse = await request(`${searchUrl}&t=pdf`);
```
[the remark this answers is not in the document]
Comment (debanjum) [reply src/interface/obsidian/src/search_modal.ts:96]: Agreed. I'll look into this soon

File: src/interface/obsidian/src/search_modal.ts [hunk 8]
```diff
@@ -138,7 +151,7 @@ export class KhojSearchModal extends SuggestModal<SearchResult> {
 
         // Open vault file at heading of chosen search result
         if (file_match) {
-            let resultHeading = result.entry.split('\n', 1)[0];
+            let resultHeading = file_match.extension !== 'pdf' ? result.entry.split('\n', 1)[0] : '';
```
[the remark this answers is not in the document]
Comment (debanjum) [reply src/interface/obsidian/src/search_modal.ts:154]: The `resultHeading` here is used to jump to the specific section within a markdown file. We do not have requirement to do that in PDF files. I do need to look into jumping to specific PDF pages when results are selected from the search modal. Keeping it simple for now, where selecting the PDF result in the search modal opens the PDF file but doesn't jump to specific page in the PDF

File: src/khoj/configure.py [hunk 9]
```diff
@@ -132,6 +133,18 @@ def configure_search(model: SearchModels, config: FullConfig, regenerate: bool,
...
 
+    # Initialize PDF Search
+    if (t == state.SearchType.Pdf or t == None) and config.content_type.pdf:
+        logger.info("💸 Setting up search for pdf")
+        # Extract Entries, Generate PDF Embeddings
+        model.pdf_search = text_search.setup(
+            PdfToJsonl,
+            config.content_type.pdf,
+            search_config=config.search_type.asymmetric,
+            regenerate=regenerate,
```
[the remark this answers is not in the document]
Comment (debanjum) [reply src/khoj/configure.py:144]: This is useful for when new PDF documents are added to the PDF index directory. We'd only want to compute embeddings for the new PDF files added, not regenerate them for the old ones by default

File: src/khoj/interface/web/index.html [hunk 10]
```diff
@@ -44,6 +44,15 @@
             }).join("\n") + `</div>`;
         }
 
+        function render_pdf(query, data) {
+            return `<div id="results-pdf">` + data.map(function (item) {
+                let compiled_lines = item.additional.compiled.split("\n");
```
[the remark this answers is not in the document]
Comment (debanjum) [reply src/khoj/interface/web/index.html:49]: Yes, the first line in the `item.additional.compiled` snippet contains the filename from which the PDF result originates. We want to render that with a separate styling

File: src/khoj/processor/conversation/gpt.py (hunk 1 above)
Comment (debanjum) [reply src/khoj/processor/conversation/gpt.py:91]: These values are used to dynamically populate the prompt with examples for GPT to get a better sense of how we expect it to extract questions. These specific variables are used for a q&q example where we ask GPT the age difference between Bob and Tom and it needs to ask multiple (2) questions "What is Tom's age", "What is Bob's age" (and later form a response based on the search results). See `prompts.py:extract_questions` for usage

File: src/khoj/processor/pdf/pdf_to_jsonl.py (hunk 5 above)
Comment (debanjum) [reply src/khoj/processor/pdf/pdf_to_jsonl.py:114]: Yes 🙂

**Comment by debanjum** [comment]
The khoj.el build failure seems unrelated to the lates pushed changes as they didn't touch khoj.el or the emacs section at all

**Pull request merged**